src/t_hash.c

```diff
@@ -922,3 +922,150 @@ void hscanCommand(client *c) {
+
+    hashSize = hashTypeLength(hash);
+    if(entryCount > hashSize)
+        entryCount = hashSize;
+    addReplyMapLen(c, entryCount);
+    hi = hashTypeInitIterator(hash);
+
+    /* if the encoding is ziplist, to only iterate once, every time
+     * we try to pick a member, the probability we pick it is the 
+     * quotient of the count left we want to pick and the count still
+     * we haven't visited in the dict, this way, we could make every member
+     * be equally picked.*/
+    if(hash->encoding == OBJ_ENCODING_ZIPLIST) {
+        while (hashTypeNext(hi) != C_ERR && entryCount != 0) {
+            randomDouble = ((double)rand()) / RAND_MAX;
+            threshold = ((double)entryCount) / (hashSize - index);
+            if(randomDouble < threshold){
+                entryCount--;
+                addHashIteratorCursorToReply(c, hi, OBJ_HASH_KEY);
+                addHashIteratorCursorToReply(c, hi, OBJ_HASH_VALUE);
+            }
+            index++;
+        }
+    } else { 
+        /* if request count is more than we have, return all entries*/      
+        if(entryCount >= hashSize) {
+            while (hashTypeNext(hi) != C_ERR) {
+                addHashIteratorCursorToReply(c, hi, OBJ_HASH_KEY);
+                addHashIteratorCursorToReply(c, hi, OBJ_HASH_VALUE);
+            }
+            return;
+        }  
+
+        /* else we need a temporary dict to support our operations */
+        static dictType dt = {
+            dictSdsHash,                /* hash function */
+            NULL,                       /* key dup */
+            NULL,                       /* val dup */
+            dictSdsKeyCompare,          /* key compare */
+            NULL,                       /* key destructor */
+            NULL,                       /* val destructor */
+            NULL                        /* allow to expand */
+        };
+        d = dictCreate(&dt,NULL);    
+
+        /* CASE 2:
+         * The number of elements inside the dict is not greater than
+         * SRANDMEMBER_SUB_STRATEGY_MUL times the number of requested elements.
+         * In this case we create a dict from scratch with all the elements, and
+         * subtract random elements to reach the requested number of elements.
+         *
+         * This is done because if the number of requested elements is just
+         * a bit less than the number of elements in the set, the natural approach
+         * used into CASE 3 is highly inefficient. */
+        if(entryCount * HRANDMEMBER_SUB_STRATEGY_MUL > hashSize) {
+            /* Add all the elements into the temporary dictionary. */
+            while((hashTypeNext(hi)) != C_ERR) {
+                int ret = DICT_ERR;
+                sds key, value;
+
+                key = hashTypeCurrentFromHashTable(hi,OBJ_HASH_KEY);
+                value = hashTypeCurrentFromHashTable(hi,OBJ_HASH_VALUE);
+                ret = dictAdd(d, key, value);
+
+                serverAssert(ret == DICT_OK);
```

Comment:
```suggestion
                serverAssert(ret == DICT_OK);
```

Since d is a new dictionary, this is not necessary.